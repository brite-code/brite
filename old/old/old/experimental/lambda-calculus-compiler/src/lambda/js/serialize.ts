import * as t from '@babel/types';
import * as Immutable from 'immutable';

import {Term, TermType} from '../term';

import {Scope} from './scope';

/**
 * The result of serialization. Contains the expression to be returned with
 * some metadata.
 */
type SerializedTerm = {
  /**
   * Should the expression be evaluated strictly? If false then the expression
   * may be ordered anywhere in our statements.
   */
  readonly strict: boolean;
  /**
   * The expression which serialization evaluates to.
   */
  readonly expression: t.Expression;
};

/**
 * Serializes a lambda calculus term to a JavaScript expression in the provided
 * scope stack.
 *
 * Returns a function which builds the resulting JavaScript expression. The
 * function should be called precisely at the place where it should be executed.
 * This way calling `serialize()` adds statements then calling the result of
 * `serialize` adds the resulting expression.
 */
function serializeTerm(
  scope: Scope,
  term: Term,
): {
  readonly statements: Immutable.Iterable<number, t.Statement>;
  readonly term: SerializedTerm;
} {
  switch (term.type) {
    case TermType.Variable: {
      return {
        statements: Immutable.List(),
        term: {
          strict: false,
          expression: scope.resolve(term.index),
        },
      };
    }

    case TermType.Abstraction: {
      return scope.block(() =>
        scope.binding(term.parameter, parameter => {
          const result = serializeTermWithStatementOutput(
            scope,
            term.body,
            expression => t.returnStatement(expression),
          );
          let body: t.Expression | t.BlockStatement;
          if (result.term !== undefined) {
            if (result.statements.isEmpty()) {
              body = result.term.expression;
            } else {
              body = t.blockStatement(
                result.statements
                  .concat(t.returnStatement(result.term.expression))
                  .toArray(),
              );
            }
          } else {
            body = t.blockStatement(result.statements.toArray());
          }
          return {
            statements: Immutable.List<t.Statement>(),
            term: {
              strict: false,
              expression: t.arrowFunctionExpression([parameter], body),
            },
          };
        }),
      );
    }

    case TermType.Application: {
      // If we are applying an abstraction then serialize a binding to avoid a
      // function call.
      if (term.callee.type === TermType.Abstraction) {
        const callee = term.callee;
        const argument = serializeTerm(scope, term.argument);
        return scope.binding(callee.parameter, parameter => {
          const bindingStatement = t.variableDeclaration('const', [
            t.variableDeclarator(parameter, argument.term.expression),
          ]);
          const body = serializeTerm(scope, callee.body);
          const statements = argument.statements.concat(
            bindingStatement,
            body.statements,
          );
          return {
            statements,
            term: body.term,
          };
        });
      } else {
        const callee = serializeTerm(scope, term.callee);
        const argument = serializeTerm(scope, term.argument);
        // If the callee needs to be evaluated strictly and the argument adds
        // some statements, then we need to hoist the callee expression above
        // the argument statements.
        if (callee.term.strict && !argument.statements.isEmpty()) {
          const variable = scope.newInternalVariable();
          const statement = t.variableDeclaration('const', [
            t.variableDeclarator(variable, callee.term.expression),
          ]);
          const statements = callee.statements.concat(
            statement,
            argument.statements,
          );
          return {
            statements,
            term: {
              strict: true,
              expression: t.callExpression(variable, [
                argument.term.expression,
              ]),
            },
          };
        } else {
          return {
            statements: callee.statements.concat(argument.statements),
            term: {
              strict: true,
              expression: t.callExpression(callee.term.expression, [
                argument.term.expression,
              ]),
            },
          };
        }
      }
    }

    // If a conditional is being serialized and we donâ€™t have a custom statement
    // output for if-statements then we might need a phi variable.
    case TermType.Conditional: {
      const phi = scope.newVariable('$phi');
      const out = (expression: t.Expression) =>
        t.expressionStatement(t.assignmentExpression('=', phi, expression));
      const test = serializeTerm(scope, term.test);
      const left = scope.block(() =>
        serializeTermWithStatementOutput(scope, term.consequent, out),
      );
      const right = scope.block(() =>
        serializeTermWithStatementOutput(scope, term.alternate, out),
      );
      if (
        !left.statements.isEmpty() ||
        !right.statements.isEmpty() ||
        left.term === undefined ||
        right.term === undefined
      ) {
        const statements = test.statements.concat(
          t.variableDeclaration('let', [t.variableDeclarator(phi)]),
          t.ifStatement(
            test.term.expression,
            t.blockStatement(
              (left.term === undefined
                ? left.statements
                : left.statements.concat(out(left.term.expression))
              ).toArray(),
            ),
            t.blockStatement(
              (right.term === undefined
                ? right.statements
                : right.statements.concat(out(right.term.expression))
              ).toArray(),
            ),
          ),
        );
        return {
          statements,
          term: {
            strict: false,
            expression: phi,
          },
        };
      } else {
        return {
          statements: test.statements,
          term: {
            strict: test.term.strict || left.term.strict || right.term.strict,
            expression: t.conditionalExpression(
              test.term.expression,
              left.term.expression,
              right.term.expression,
            ),
          },
        };
      }
    }

    // Serialize native terms by serializing all the inputs of the term and then
    // calling the JavaScript serializer function for the native term.
    case TermType.Native: {
      return combine(
        scope,
        term.inputs.map(input => serializeTerm(scope, input)),
        inputs => {
          let strict = false;
          const serializeInputs = inputs.map(input => {
            strict = strict || input.strict;
            return input.expression;
          });
          return {
            strict,
            expression: term.serializers.js(serializeInputs),
          };
        },
      );
    }
  }
}

/**
 * Serializes a term with a custom output function for returning expressions
 * from the statement. For instance, perhaps the last expression should
 * be returned.
 */
function serializeTermWithStatementOutput(
  scope: Scope,
  term: Term,
  out: (expression: t.Expression) => t.Statement,
): {
  readonly statements: Immutable.Iterable<number, t.Statement>;
  readonly term: SerializedTerm | undefined;
} {
  // Special case bindings. The body of a binding should be serialized with our
  // custom statement output function.
  if (
    term.type === TermType.Application &&
    term.callee.type === TermType.Abstraction
  ) {
    const callee = term.callee;
    const argument = serializeTerm(scope, term.argument);
    return scope.binding(callee.parameter, parameter => {
      const bindingStatement = t.variableDeclaration('const', [
        t.variableDeclarator(parameter, argument.term.expression),
      ]);
      const body = serializeTermWithStatementOutput(scope, callee.body, out);
      const statements = argument.statements.concat(
        bindingStatement,
        body.statements,
      );
      return {
        statements,
        term: body.term,
      };
    });
  }

  // Special case conditionals so that the last statement in each conditional
  // branch may use the statement output function.
  if (term.type === TermType.Conditional) {
    const test = serializeTerm(scope, term.test);
    const left = scope.block(() =>
      serializeTermWithStatementOutput(scope, term.consequent, out),
    );
    const right = scope.block(() =>
      serializeTermWithStatementOutput(scope, term.alternate, out),
    );
    if (
      !left.statements.isEmpty() ||
      !right.statements.isEmpty() ||
      left.term === undefined ||
      right.term === undefined
    ) {
      const statements = test.statements.concat(
        t.ifStatement(
          test.term.expression,
          t.blockStatement(
            (left.term === undefined
              ? left.statements
              : left.statements.concat(out(left.term.expression))
            ).toArray(),
          ),
          t.blockStatement(
            (right.term === undefined
              ? right.statements
              : right.statements.concat(out(right.term.expression))
            ).toArray(),
          ),
        ),
      );
      return {
        statements,
        term: undefined,
      };
    } else {
      return {
        statements: test.statements,
        term: {
          strict: test.term.strict || left.term.strict || right.term.strict,
          expression: t.conditionalExpression(
            test.term.expression,
            left.term.expression,
            right.term.expression,
          ),
        },
      };
    }
  }

  return serializeTerm(scope, term);
}

/**
 * Serializes a lambda calculus term to a JavaScript program.
 */
export function serialize(term: Term): t.Program {
  const result = serializeTerm(new Scope(), term);
  const programStatements = result.statements.toArray();
  programStatements.push(t.expressionStatement(result.term.expression));
  return t.program(programStatements);
}

/**
 * Combines an array of serialized terms into a single serialized term with a
 * combiner function. Concatenates all the statement lists together and
 * maintains the ordering of strict expressions.
 */
function combine(
  scope: Scope,
  results: ReadonlyArray<{
    readonly statements: Immutable.Iterable<number, t.Statement>;
    readonly term: SerializedTerm;
  }>,
  combiner: (expressions: ReadonlyArray<SerializedTerm>) => SerializedTerm,
): {
  readonly statements: Immutable.Iterable<number, t.Statement>;
  readonly term: SerializedTerm;
} {
  // Create the initial variables we will be modifying.
  let statements: Immutable.Iterable<number, t.Statement> = Immutable.List();
  const terms = Array(results.length);
  // Iterate through the results we were provided backwards. This way we can
  // detect if there are statements _after_ the position of a strict expression.
  // If there are then we need to hoist the expression into a statement.
  for (let i = results.length - 1; i >= 0; i--) {
    const result = results[i];
    // If our result expression is strictly evaluated and there are some
    // statements evaluated after our strict expression then we need to take
    // care to evaluate our expression in the proper order. So we hoist our
    // strict expression into a variable declaration so that it evaluates after
    // its own statements and before the other statements. We then use a
    // temporary identifier for the hoisted expression.
    if (result.term.strict && !statements.isEmpty()) {
      const identifier = scope.newInternalVariable();
      terms[i] = {strict: false, expression: identifier};
      const bindingStatement = t.variableDeclaration('const', [
        t.variableDeclarator(identifier, result.term.expression),
      ]);
      statements = result.statements.concat(bindingStatement, statements);
    } else {
      // Add our expression to our array and prepend our statements to the list.
      terms[i] = result.term;
      statements = result.statements.concat(statements);
    }
  }
  return {
    statements,
    term: combiner(terms),
  };
}
