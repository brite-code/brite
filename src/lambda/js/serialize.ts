import * as t from '@babel/types';
import {Term, TermType} from '../term';
import {ScopeStack} from './scope';

type JsTerm = Term<t.Expression>;

/**
 * Serializes a lambda calculus term to a JavaScript expression in the provided
 * scope stack.
 */
function serialize(term: JsTerm, scope: ScopeStack): t.Expression {
  switch (term.type) {
    case TermType.Variable: {
      const variable = scope.resolveVariable(term.name);
      if (variable === undefined) {
        throw new Error(`Could not resolve variable "${term.name}"`);
      }
      return variable;
    }
    case TermType.Abstraction: {
      const [{parameter, result}, statements] = scope.nest(() => {
        const parameter = scope.declareVariable(term.parameter);
        const result = serialize(term.body, scope);
        return {parameter, result};
      });
      let body;
      if (statements.length === 0) {
        body = result;
      } else {
        statements.push(t.returnStatement(result));
        body = t.blockStatement(statements);
      }
      return t.arrowFunctionExpression([parameter], body);
    }
    case TermType.Application: {
      // If we are applying an abstraction then serialize a binding to avoid a
      // function call.
      if (term.callee.type === TermType.Abstraction) {
        const init = serialize(term.argument, scope);
        const parameter = scope.declareVariable(term.callee.parameter);
        scope.addStatement(
          t.variableDeclaration('const', [
            t.variableDeclarator(parameter, init),
          ]),
        );
        return serialize(term.callee.body, scope);
      } else {
        return t.callExpression(serialize(term.callee, scope), [
          serialize(term.argument, scope),
        ]);
      }
    }
    case TermType.Native: {
      return term.serialize(term.variables.map(name => t.identifier(name)));
    }
  }
}

/**
 * Serializes a lambda calculus term to a JavaScript expression.
 */
function serializeStart(term: JsTerm): t.Program {
  const [result, statements] = ScopeStack.create(scope =>
    serialize(term, scope),
  );
  statements.push(t.expressionStatement(result));
  return t.program(statements);
}

export {serializeStart as serialize};
