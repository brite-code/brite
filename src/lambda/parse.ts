import * as Immutable from 'immutable';

import {
  Identifier,
  Term,
  abstraction,
  application,
  binding,
  variable,
} from './term';

/**
 * Parses a lambda calculus term.
 *
 * - `x` → `variable(x)`
 * - `(λx.y)` → `abstraction(x, y)`
 * - `(x y)` → `application(x, y)`
 * - `let x = y in z` → `binding(x, y, z)`
 */
export function parse<T>(source: string): Term<T> {
  const iterator = peekable(tokenize(source[Symbol.iterator]()));
  const term = parseTerm<T>(iterator, Scope.empty());
  const step = iterator.next();
  if (!step.done) {
    throw new Error(`Unexpected token "${step.value.type}" expected ending`);
  }
  return term;
}

/**
 * Does the actual lambda calculus parsing on an iterator.
 */
function parseTerm<T>(
  iterator: PeekableIterator<Token>,
  scope: Scope,
): Term<T> {
  // Parse an abstraction.
  if (tryParseToken(iterator, TokenType.Lambda)) {
    const [newScope, name] = Scope.declare(scope, parseIdentifier(iterator));
    parseToken(iterator, TokenType.Dot);
    const body = parseTerm<T>(iterator, newScope);
    return abstraction(name, body);
  }

  // Parse a binding
  if (tryParseToken(iterator, TokenType.Let)) {
    const [newScope, name] = Scope.declare(scope, parseIdentifier(iterator));
    parseToken(iterator, TokenType.Equals);
    const value = parseTerm<T>(iterator, scope);
    parseToken(iterator, TokenType.In);
    const body = parseTerm<T>(iterator, newScope);
    return binding<T>(name, value, body);
  }

  // Parses an unwrapped term.
  let term = parseUnwrappedTerm<T>(iterator, scope);
  if (term === undefined) {
    const step = iterator.next();
    if (step.done) throw new Error('Unexpected ending expected term');
    throw new Error(`Unexpected token "${step.value.type}" expected term`);
  }

  // Parse as many application arguments as we can.
  while (true) {
    const argument = parseUnwrappedTerm<T>(iterator, scope);
    if (argument === undefined) {
      break;
    } else {
      term = application(term, argument);
    }
  }

  return term;
}

/**
 * Parses an unwrapped term. Either a variable or a wrapped term. Returns
 * undefined if a term could not be parsed.
 */
function parseUnwrappedTerm<T>(
  iterator: PeekableIterator<Token>,
  scope: Scope,
): Term<T> | undefined {
  // Parse a term inside parentheses.
  if (tryParseToken(iterator, TokenType.ParenLeft)) {
    const term = parseTerm<T>(iterator, scope);
    parseToken(iterator, TokenType.ParenRight);
    return term;
  }

  // Parse a variable.
  const step = iterator.peek();
  if (!step.done && step.value.type === TokenType.Identifier) {
    iterator.next();
    return variable(Scope.resolve(scope, step.value.data));
  }

  // Otherwise we couldn’t parse any unwrapped terms.
  return undefined;
}

/**
 * Parses a token in the iterator and throws if it was not found.
 */
function parseToken(iterator: PeekableIterator<Token>, token: TokenType) {
  const step = iterator.next();
  if (step.done) {
    throw new Error(`Unexpected ending expected "${token}"`);
  }
  if (step.value.type !== token) {
    throw new Error(
      `Unexpected token "${step.value.type}" expected token "${token}"`,
    );
  }
}

/**
 * Tries to parse a token of the provided type. Returns true if it was found and
 * false if it was not.
 */
function tryParseToken(
  iterator: PeekableIterator<Token>,
  token: TokenType,
): boolean {
  const step = iterator.peek();
  const found = !step.done && step.value.type === token;
  if (found) iterator.next();
  return found;
}

/**
 * Parses an identifier from the iterator. Throws if no identifier was found.
 */
function parseIdentifier(iterator: PeekableIterator<Token>): string {
  const step = iterator.next();
  if (step.done) {
    throw new Error('Unexpected ending expected identifier');
  }
  if (step.value.type !== TokenType.Identifier) {
    throw new Error(
      `Unexpected token "${step.value.type}" expected identifier`,
    );
  }
  return step.value.data;
}

type Scope = {
  readonly identifiers: Array<string>;
  readonly variables: Immutable.Map<string, Identifier>;
};

namespace Scope {
  export function empty(): Scope {
    return {identifiers: [], variables: Immutable.Map()};
  }

  export function declare(scope: Scope, name: string): [Scope, Identifier] {
    const identifier = scope.identifiers.length;
    scope.identifiers.push(name);
    const newScope = {
      identifiers: scope.identifiers,
      variables: scope.variables.set(name, identifier),
    };
    return [newScope, identifier];
  }

  export function resolve<T>(scope: Scope, name: string): Identifier {
    const identifier = scope.variables.get(name);
    if (identifier === undefined) {
      throw new Error(`Variable "${name}" was not declared.`);
    }
    return identifier;
  }
}

/**
 * The type of a token.
 */
const enum TokenType {
  Lambda = 'λ',
  Dot = '.',
  ParenLeft = '(',
  ParenRight = ')',
  Equals = '=',
  Let = 'let',
  In = 'in',
  Identifier = 'identifier',
}

/**
 * A token value. Mostly the same as `TokenType` except that some types are
 * associated with a value. Like `TokenType.Identifier`.
 */
type Token =
  | {type: TokenType.Lambda}
  | {type: TokenType.Dot}
  | {type: TokenType.ParenLeft}
  | {type: TokenType.ParenRight}
  | {type: TokenType.Equals}
  | {type: TokenType.Let}
  | {type: TokenType.In}
  | {type: TokenType.Identifier; data: string};

const identifierStart = /\w/;
const identifierContinue = /[\w\d]/;
const whitespace = /\s/;

/**
 * Takes a source iterator of characters and returns an iterator of tokens. The
 * token iterator is much easier to work with.
 */
function* tokenize(source: Iterator<string>): IterableIterator<Token> {
  let step = source.next();
  let identifier: string | undefined = undefined;
  while (true) {
    // If an identifier has been started either add to the identifier or yield
    // the completed identifier or keyword.
    if (identifier !== undefined) {
      if (!step.done && identifierContinue.test(step.value)) {
        identifier += step.value;
        step = source.next();
        continue;
      } else {
        if (identifier === 'let') {
          yield {type: TokenType.Let};
        } else if (identifier === 'in') {
          yield {type: TokenType.In};
        } else {
          yield {type: TokenType.Identifier, data: identifier};
        }
        identifier = undefined;
      }
    }
    // If we are done then break out of the loop!
    if (step.done) {
      break;
    }
    if (step.value === 'λ') {
      yield {type: TokenType.Lambda};
    } else if (step.value === '.') {
      yield {type: TokenType.Dot};
    } else if (step.value === '(') {
      yield {type: TokenType.ParenLeft};
    } else if (step.value === ')') {
      yield {type: TokenType.ParenRight};
    } else if (step.value === '=') {
      yield {type: TokenType.Equals};
    } else if (identifierStart.test(step.value)) {
      // Start an identifier with this step.
      identifier = step.value;
    } else if (whitespace.test(step.value)) {
      // noop
    } else {
      throw new Error(`Unexpected character "${step.value}"`);
    }
    step = source.next();
  }
}

/**
 * A peekable iterator allows you to peek the next item without consuming it.
 * Note that peekable iterators do not allow you to pass in a value with
 * `next()`. The interface of `PeekableIterator<T>` is much more limited then
 * the interface of `Iterator<T>` to allow for peeking.
 */
interface PeekableIterator<T> {
  next(): IteratorResult<T>;
  peek(): IteratorResult<T>;
}

/**
 * Turns an iterator into a peekable iterator.
 */
function peekable<T>(iterator: Iterator<T>): PeekableIterator<T> {
  let peeking: IteratorResult<T> | undefined = undefined;
  return {
    next: () => {
      if (peeking !== undefined) {
        const next = peeking;
        peeking = undefined;
        return next;
      } else {
        return iterator.next();
      }
    },
    peek: () => {
      if (peeking === undefined) {
        peeking = iterator.next();
      }
      return peeking;
    },
  };
}
