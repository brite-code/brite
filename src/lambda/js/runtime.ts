import * as t from '@babel/types';
import {
  Term,
  abstraction,
  native,
  application,
  getFreeVariables,
} from '../term';
import {parse} from '../parse';

type JsTerm = Term<t.Expression>;

/**
 * Adds the applicable runtime to some lambda calculus term.
 */
export function addRuntime(initialTerm: JsTerm): JsTerm {
  const free = getFreeVariables(initialTerm);
  const nativeVariables = getNativeVariables();
  let term = initialTerm;
  for (const [name, value] of nativeVariables) {
    if (!free.has(name)) continue;
    // TODO: This isn’t very efficient. Preferably we would provide our native
    // variables directly to a compiler which would pull them in on demand.
    term = application(abstraction(name, term), value);
  }
  return term;
}

let nativeVariables: Map<string, JsTerm>;

/**
 * Gets the native variables for our programs.
 *
 * If this function has not been called before we initialize the
 * native variables.
 */
function getNativeVariables(): Map<string, JsTerm> {
  if (nativeVariables === undefined) {
    nativeVariables = initNativeVariables();
  }
  return nativeVariables;
}

/**
 * Initializes the native variables for our programs.
 */
function initNativeVariables(): Map<string, JsTerm> {
  const variables = new Map<string, JsTerm>();

  // Identity function
  variables.set('id', parse('λx.x'));

  // Fix-point combinator (Z-combinator) since we are strictly evaluated.
  // https://en.wikipedia.org/wiki/Fixed-point_combinator
  variables.set('fix', parse('λf.(λx.λv.((f (x x)) v)) (λx.λv.((f (x x)) v))'));

  // Common numbers
  variables.set('zero', native([], (): t.Expression => t.numericLiteral(0)));
  variables.set('one', native([], (): t.Expression => t.numericLiteral(1)));

  // Addition
  const add: JsTerm = native(['x', 'y'], ([x, y]) =>
    t.binaryExpression('+', x, y),
  );
  variables.set('add', abstraction('x', abstraction('y', add)));

  // Subtraction
  const sub: JsTerm = native(['x', 'y'], ([x, y]) =>
    t.binaryExpression('-', x, y),
  );
  variables.set('sub', abstraction('x', abstraction('y', sub)));

  // Multiplication
  const mul: JsTerm = native(['x', 'y'], ([x, y]) =>
    t.binaryExpression('*', x, y),
  );
  variables.set('mul', abstraction('x', abstraction('y', mul)));

  // Division
  const div: JsTerm = native(['x', 'y'], ([x, y]) =>
    t.binaryExpression('/', x, y),
  );
  variables.set('div', abstraction('x', abstraction('y', div)));

  return variables;
}
