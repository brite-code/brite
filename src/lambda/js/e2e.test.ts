// tslint:disable no-eval

import generate from '@babel/generator';

import {parse} from '../parse';

import {addRuntime} from './runtime';
import {serialize} from './serialize';

[
  {
    source: 'fix (λf.λn.if (eq n zero) (const one) (λz.mul n (f (sub n one))))',
    cases: [[0, 1], [1, 1], [2, 2], [3, 6], [4, 24], [5, 120]],
  },
  {
    source:
      'fix (λf.λn.if (eq n zero) (const zero) (λz.if (eq n one) (const one) (λz.add (f (sub n one)) (f (sub (sub n one) one)))))',
    cases: [[0, 0], [1, 1], [2, 1], [3, 2], [4, 3], [5, 5], [6, 8], [7, 13]],
  },
  {
    source:
      'let fix = λf.(λx.f (λv.x x v)) (λx.f (λv.x x v)) in fix (λf.λn.if (eq n zero) (const one) (λz.mul n (f (sub n one))))',
    cases: [[0, 1], [1, 1], [2, 2], [3, 6], [4, 24], [5, 120]],
  },
].forEach(({source, cases}) => {
  test(source, () => {
    const {code} = generate(serialize(addRuntime(parse(source))));
    expect(code).toMatchSnapshot();
    const f = eval(code);
    cases.forEach(([input, output]) => {
      expect(f(input)).toEqual(output);
    });
  });
});
