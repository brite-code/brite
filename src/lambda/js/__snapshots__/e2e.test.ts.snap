// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fix (λf.λn.if (eq n zero) (const one) (λz.mul n (f (sub n one)))) 1`] = `
"const eq = x => y => x === y;

const if_ = x => y => z => x ? y() : z();

const mul = x => y => x * y;

const sub = x => y => x - y;

const one = 1;
const zero = 0;

const fix = f => {
  const x = x => f(v => x(x)(v));

  return f(v => x(x)(v));
};

const const_ = x => y => x;

fix(f => n => if_(eq(n)(zero))(const_(one))(z => mul(n)(f(sub(n)(one)))));"
`;

exports[`fix (λf.λn.if (eq n zero) (const zero) (λz.if (eq n one) (const one) (λz.add (f (sub n one)) (f (sub (sub n one) one))))) 1`] = `
"const eq = x => y => x === y;

const if_ = x => y => z => x ? y() : z();

const sub = x => y => x - y;

const add = x => y => x + y;

const one = 1;
const zero = 0;

const fix = f => {
  const x = x => f(v => x(x)(v));

  return f(v => x(x)(v));
};

const const_ = x => y => x;

fix(f => n => if_(eq(n)(zero))(const_(zero))(z => if_(eq(n)(one))(const_(one))(z => add(f(sub(n)(one)))(f(sub(sub(n)(one))(one))))));"
`;
